### PR TITLE
cache ttl variant

### PR DESCRIPTION
**Pull request by djabarovgeorge**
### What change does this PR introduce? 

adds cache TTL variant

### Why was this change needed?

after talk with @Cliftonz, he suggested to add TTL variant in order to reduce potential load on the Redis

### CODE REVIEW

**Comment by Cliftonz** [comment]
@djabarovgeorge  I would recommend the invariant be +/- instead of just plus. This way the distribution is centered around the TTL instead of TTL + ((TTL *percentage)/2)
This way the users know that the average TTL  will be around the number they set if they measure it.

You should be able to accomplish this by the following:

`
return (num - (variant /2)) + (num * variant);
`

File: apps/api/src/app/shared/services/cache/cache.service.ts [hunk 1]
```diff
@@ -98,7 +99,14 @@ export class CacheService implements ICacheService {
   private updateTtl(key: string, options?: CachingConfig) {
     const seconds = options?.ttl || this.cacheTtl;
 
-    return this.client.expire(key, seconds);
+    return this.client.expire(key, this.ttlVariant(seconds));
+  }
+
+  private ttlVariant(num) {
+    const randomNum = Math.random();
+    const variant = this.TTL_VARIANT_PERCENTAGE * randomNum;
```
Comment (Cliftonz) [inline apps/api/src/app/shared/services/cache/cache.service.ts:107]: The variant should be,
`
const variant = this.TTL_VARIANT_PERCENTAGE * num;
`
The current equation will make the variance smaller by a random number.

**Comment by Cliftonz** [comment]
@djabarovgeorge My apologies for the misleading messages before. Here are two alternatives to do this will make the normal distribution centered around the actual TTL set by the user.
`
const variant = (this.TTL_VARIANT_PERCENTAGE / 2) * num * math.random();
if ( Math.random() < 0.5){
    return num + variant;
} else {
    return num - variant;
}
`
another alternative that is branchless is:
`
const variant = this.TTL_VARIANT_PERCENTAGE * num * math.random();
return num -  ((this.TTL_VARIANT_PERCENTAGE * num)/2) + variant
`
The second one will be more performant while the first one is easier to read.

Let me know if you have any questions.

**Comment by scopsy** [comment]
Could you elaborate on how this can help with load?

**Comment by Cliftonz** [comment]
The problem we are trying to prevent is called the Thundering Herd Problem.

Assuming the system is under load and is hitting a sufficient amount of data in the cache, then if the set of most accessed data expires then every call for that data will go to the database and we will see spikes in load across the system.

In a HA environment, this may cause the system to incorrectly scale and/or not meet SLO for response times or at worst deadlock and fail.

By adding variance to the TTL, we can minimize the number of calls to the database at the same time when data expires in the cache.

Another mitigation that we will need to implement is Request Collapsing. This ensures that only one request to the database for a piece of data is done at any one instance. 

load

**Comment by Cliftonz** [comment]
Here is a good resource to read about this problem
https://developer.fastly.com/learning/concepts/request-collapsing/#without-request-collapsing---the-cache-stampede-problem

As they say, the two hardest things are caching invalidation and naming.

**Pull request merged**